core/src/main/scala/org/apache/gearpump/util/Graph.scala

```diff
@@ -243,13 +259,34 @@ class Graph[N, E](vertexList: List[N], edgeList: List[(N, E, N)]) extends Serial
    * The node returned by Iterator is stable sorted.
    */
   def topologicalOrderIterator: Iterator[N] = {
-    val newGraph = copy
-    var output = List.empty[N]
-
-    while (!newGraph.isEmpty) {
-      output ++= newGraph.removeZeroInDegree
+    tryTopologicalOrderIterator.get
```

Comment:
This will throw exception on failure ? The callers don't know.

Reply:
I think exception is better than dead loop

Reply:
Then return a `Try` is better since the caller will be aware this method may fail then.